Contribute the handwritten-japanese-recognition-0001 model

cocokale opened the pull request:
Contribute the handwritten-japanese-recognition-0001 model along with the accuracy checker tool support for it, and a python demo for it

vladimir-dudnik commented:
Jenkins please retry a build

cocokale commented:
> you do not need model.yml for intel models. it will be generated automatically by release infrastructure, please remove this file.

Hi Vladimir, removal finished. :)

IRDonch commented:
Question: couldn't the existing `text_detection_demo` with some minor updates handle this model? It looks like its interface is more-or-less the same as the one of `handwritten-score-recognition-0003`, which is supported by `text_detection_demo`.

eaidova commented on tools/accuracy_checker/accuracy_checker/metrics/character_recognition.py:
```diff
@@ -37,3 +37,22 @@ def evaluate(self, annotations, predictions):
+    prediction_types = (CharacterRecognitionPrediction, )
+
+    def configure(self):
+        self.accuracy = AverageEditdistanceMeter(lambda annotation, prediction: editdistance.eval(prediction, annotation))
```
probably, if it is only one metric which use `AverageEditdistanceMeter` you can use default realization without lambda. Just make it the default loss

eaidova commented on tools/accuracy_checker/accuracy_checker/metrics/average_editdistance_meter.py:
```diff
@@ -0,0 +1,41 @@
+import editdistance
+
+class AverageEditdistanceMeter:
+    def __init__(self, loss=None, counter=None):
```
I do not see the counter usage

eaidova commented on tools/accuracy_checker/accuracy_checker/metrics/average_editdistance_meter.py:
```diff
@@ -0,0 +1,41 @@
+limitations under the License.
+"""
+
+import editdistance
```
You introduce the new dependency. Is it necessary?
If yes, please update `requirements.in`

eaidova commented on tools/accuracy_checker/accuracy_checker/adapters/text_detection.py:
```diff
@@ -998,3 +998,72 @@ def restore_rectangle(origin, geometry):
+        parameters = super().parameters()
+        parameters.update({
+            'blank_label': NumberField(
+                optional=True, value_type=int, min_value=0, description="Index of the CTC blank label."
```
you can set zero as default value, the code below shows that you do not allow usage without blank label at all, so you can avoid additional checks with such replacement
```suggestion
                optional=True, value_type=int, min_value=0, default=0, description="Index of the CTC blank label."
```

eaidova replied:
BTW, why is ctc_beam_search_decoder with beam_size=1 and blank_label=0  unsuitable for you?

cocokale replied:
> BTW, why is ctc_beam_search_decoder with beam_size=1 and blank_label=0 unsuitable for you?

Hi eaidova,
For our model, the class number (the dimension D is 1161) is much bigger than the english text recognition (D = 37). If our model uses the existing beam search decoder, the running will be very slow and hang... So we just use numpy.argmax to select the max index in the D dimension. The results on our side show for our model and task, the approach is effective and efficient enough.

eaidova replied:
ok, thanks for explanation

eaidova commented on tools/accuracy_checker/accuracy_checker/preprocessor/geometric_transformations.py:
```diff
@@ -380,6 +381,9 @@ def parameters(cls):
                 optional=True, default=False, description="Allow to use numpy for padding instead default OpenCV."
+            ),
+            'numpy_pad_mode': StringField(
+                optional=True, default='constant', description="If use_numpy is True, Numpy padding mode,including constant, edge, mean, etc."
```
it makes sense to have the list of possible values for this parameter.
please specify it with `choices`  argument

eaidova commented on tools/accuracy_checker/configs/handwritten-japanese-recognition-0001.yml:
```diff
@@ -0,0 +1,40 @@
+            interpolation: AREA
+            aspect_ratio_scale: width
+            size: 96
+          - type: normalization
```
usually, we place normalization to model IR during MO conversion

vladimir-dudnik replied:
moreover, for intel models it is requirement

eaidova commented on tools/accuracy_checker/configs/handwritten-japanese-recognition-0001.yml:
```diff
@@ -0,0 +1,40 @@
+          - type: label_level_recognition_accuracy
+
+global_definitions: ../dataset_definitions.yml
+
```
please remove extra empty line

eaidova commented on tools/accuracy_checker/configs/handwritten-japanese-recognition-0001.yml:
```diff
@@ -0,0 +1,40 @@
+
+    launchers:
+      - framework: dlsdk
+        device: cpu
```
please remove device, you can set it via `--target_device` command line argument

eaidova commented on tools/accuracy_checker/configs/handwritten-japanese-recognition-0001.yml:
```diff
@@ -0,0 +1,40 @@
+      - framework: dlsdk
+        device: cpu
+        tags:
+          - FP16
```
@vladimir-dudnik @IRDonch will the model available only in FP16 precision or FP32 version also be places in OMZ?

vladimir-dudnik replied:
@eaidova @cocokale I did not saw the model itself yet, but I think it definitely have to be available in FP32 precision as well.

eaidova commented on tools/accuracy_checker/configs/handwritten-japanese-recognition-0001.yml:
```diff
@@ -0,0 +1,40 @@
+        tags:
+          - FP16
+        model:   intel/handwritten-japanese-recognition-0001.xml
+        weights: intel/handwritten-japanese-recognition-0001.bin
```
```suggestion
        model:   intel/handwritten-japanese-recognition-0001/FP16/handwritten-japanese-recognition-0001.xml
        weights: intel/handwritten-japanese-recognition-0001/FP16/handwritten-japanese-recognition-0001.bin
```

vladimir-dudnik commented:
@cocokale please resolve conflicts

eaidova commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,137 @@
+    rsz = cv2.resize(src, (tw, th), interpolation=cv2.INTER_AREA).astype(np.float32)
+
+    # Normalize the image to [-1, 1]
+    rsz = (rsz - 127.5) / 127.5
```
here normalization also should be removed

eaidova commented on tools/accuracy_checker/accuracy_checker/preprocessor/geometric_transformations.py:
```diff
@@ -435,7 +440,7 @@ def process(self, image, annotation_meta=None):
         pad = self.pad_func(pref_width, pref_height, width, height)
         image.metadata['padding'] = pad
-        padding_realization_func = self._opencv_padding if not self.use_numpy else self._numpy_padding
+        padding_realization_func = self._opencv_padding if not self.use_numpy else partial(self._numpy_padding, mode=self.numpy_pad_mode)
```
`self._numpy_padding` is method and has access to class instance fields, So you are able directly use `self.numpy_pad_mode` inside it without providing it as argument and partial function creation.

cocokale replied:
> `self._numpy_padding` is method and has access to class instance fields, So you are able directly use `self.numpy_pad_mode` inside it without providing it as argument and partial function creation.

Hi,
The numpy_padding is called by:
image.data = padding_realization_func(image.data, pad)
here the mode is not passed in, because it also supports opencv padding, which does not have the mode parameter.

eaidova replied:
no, I mean what prevents to do something like this?
```python
    def _numpy_padding(self, image, pad):
        pad_values = (
            (self.pad_value[0], self.pad_value[0]),
            (self.pad_value[1], self.pad_value[1]),
            (self.pad_value[2], self.pad_value[2])
        )
        if self.numpy_pad_mode != 'constant':
            return np.pad(
                image, ((pad[0], pad[2]), (pad[1], pad[3]), (0, 0)),
                mode=elf.numpy_pad_mode
            )
        else:
            return np.pad(
                image, ((pad[0], pad[2]), (pad[1], pad[3]), (0, 0)),
                mode=elf.numpy_pad_mode, constant_values=pad_values
            )
```

cocokale replied:
Understand. I'll modify it.

cocokale replied:
OK, I understand. This will work using self.numpy_pad_mode

eaidova commented:
jenkins ok to test

eaidova commented:
@cocokale, Accuracy Checker tests fialed with error:
```
ImportError: No module named 'editdistance'
```
looks like you need to update https://github.com/opencv/open_model_zoo/blob/develop/ci/requirements-ac-test.txt

cocokale commented:
> @cocokale, Accuracy Checker tests fialed with error:
> 
> ```
> ImportError: No module named 'editdistance'
> ```
> 
> looks like you need to update https://github.com/opencv/open_model_zoo/blob/develop/ci/requirements-ac-test.txt

fixed

cocokale commented:
@eaidova Seems like another error has occurred...

eaidova commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,135 @@
+                           "value is CPU")
+    args.add_argument("-ni", "--number_iter", type=int, default=1,
+                      help="Optional. Number of inference iterations")
+    args.add_argument("-lang", "--lang", type=str, default='jp',
```
I can not understand this parameter appointment. Is there the difference in format of charlist which depends on language?

cocokale replied:
No. Currently the model only supports Japanese. This parameter can be deleted.

eaidova commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,135 @@
+    chrs_list_file = ''
+    if args.lang == 'jp':
+        chrs_list_file = args.charlist
+        with open(chrs_list_file, 'r') as f:
```
probably you need specify encoding for stable work on different machines

eaidova commented:
> @eaidova Seems like another error has occurred...

@cocokale Currently it is pylint errors, could you please fix them?
```
13:56:51  ************* Module accuracy_checker.adapters.text_detection
13:56:51  accuracy_checker/adapters/text_detection.py:1029: [C0326(bad-whitespace), ] Exactly one space required after comma
13:56:51          preds_index = preds_index.transpose(1,0)
13:56:51                                               ^
13:56:51  accuracy_checker/adapters/text_detection.py:1069: [C0304(missing-final-newline), ] Final newline missing
13:56:51  accuracy_checker/adapters/text_detection.py:1: [C0302(too-many-lines), ] Too many lines in module (1069/1000)
13:56:51  ************* Module accuracy_checker.annotation_converters.kondate
13:56:51  accuracy_checker/annotation_converters/kondate.py:39: [C0301(line-too-long), ] Line too long (128/120)
13:56:51  accuracy_checker/annotation_converters/kondate.py:40: [C0330(bad-continuation), ] Wrong continued indentation.
13:56:51                  )
13:56:51                  ^                              ||
13:56:51  accuracy_checker/annotation_converters/kondate.py:17: [W0611(unused-import), ] Unused numpy imported as np
13:56:51  accuracy_checker/annotation_converters/kondate.py:18: [C0411(wrong-import-order), ] standard import "from typing import Union" should be placed before "import numpy as np"
13:56:51  accuracy_checker/annotation_converters/kondate.py:19: [C0411(wrong-import-order), ] standard import "from pathlib import Path" should be placed before "import numpy as np"
13:56:51  ************* Module accuracy_checker.metrics.average_editdistance_meter
13:56:51  accuracy_checker/metrics/average_editdistance_meter.py:17: [W0611(unused-import), ] Unused import editdistance
13:56:51  ************* Module accuracy_checker.metrics.character_recognition
13:56:51  accuracy_checker/metrics/character_recognition.py:58: [C0304(missing-final-newline), ] Final newline missing
13:56:51  ************* Module accuracy_checker.preprocessor.geometric_transformations
13:56:51  accuracy_checker/preprocessor/geometric_transformations.py:414: [C0301(line-too-long), ] Line too long (220/120)
13:56:51  accuracy_checker/preprocessor/geometric_transformations.py:470: [R1705(no-else-return), Padding._numpy_padding] Unnecessary "else" after "return"
```

eaidova commented:
@cocokale please also add your model tp [index.md](https://github.com/opencv/open_model_zoo/blob/develop/models/intel/index.md)

cocokale commented:
> > @eaidova Seems like another error has occurred...
> 
> @cocokale Currently it is pylint errors, could you please fix them?
> 
> ```
> 13:56:51  ************* Module accuracy_checker.adapters.text_detection
> 13:56:51  accuracy_checker/adapters/text_detection.py:1029: [C0326(bad-whitespace), ] Exactly one space required after comma
> 13:56:51          preds_index = preds_index.transpose(1,0)
> 13:56:51                                               ^
> 13:56:51  accuracy_checker/adapters/text_detection.py:1069: [C0304(missing-final-newline), ] Final newline missing
> 13:56:51  accuracy_checker/adapters/text_detection.py:1: [C0302(too-many-lines), ] Too many lines in module (1069/1000)
> 13:56:51  ************* Module accuracy_checker.annotation_converters.kondate
> 13:56:51  accuracy_checker/annotation_converters/kondate.py:39: [C0301(line-too-long), ] Line too long (128/120)
> 13:56:51  accuracy_checker/annotation_converters/kondate.py:40: [C0330(bad-continuation), ] Wrong continued indentation.
> 13:56:51                  )
> 13:56:51                  ^                              ||
> 13:56:51  accuracy_checker/annotation_converters/kondate.py:17: [W0611(unused-import), ] Unused numpy imported as np
> 13:56:51  accuracy_checker/annotation_converters/kondate.py:18: [C0411(wrong-import-order), ] standard import "from typing import Union" should be placed before "import numpy as np"
> 13:56:51  accuracy_checker/annotation_converters/kondate.py:19: [C0411(wrong-import-order), ] standard import "from pathlib import Path" should be placed before "import numpy as np"
> 13:56:51  ************* Module accuracy_checker.metrics.average_editdistance_meter
> 13:56:51  accuracy_checker/metrics/average_editdistance_meter.py:17: [W0611(unused-import), ] Unused import editdistance
> 13:56:51  ************* Module accuracy_checker.metrics.character_recognition
> 13:56:51  accuracy_checker/metrics/character_recognition.py:58: [C0304(missing-final-newline), ] Final newline missing
> 13:56:51  ************* Module accuracy_checker.preprocessor.geometric_transformations
> 13:56:51  accuracy_checker/preprocessor/geometric_transformations.py:414: [C0301(line-too-long), ] Line too long (220/120)
> 13:56:51  accuracy_checker/preprocessor/geometric_transformations.py:470: [R1705(no-else-return), Padding._numpy_padding] Unnecessary "else" after "return"
> ```

I've fixed all the pylint errors except the too many lines error in text_detection.py... I'm not sure about what should I do. Should I put the ctcgreedysearchdecoder in another file? Or this error can just be skipped?

eaidova commented:
> Should I put the ctcgreedysearchdecoder in another file?

Yes, I think it is the best way to handle this error (or separate text detection and text recognition - move both `ctc_greedy_search_decoder` and `ctc_beam_search_decoder` to another file)

>I've fixed all the pylint errors except the too many lines error in text_detection.py...

There is also
```
14:34:25  accuracy_checker/annotation_converters/kondate.py:39: [C0330(bad-continuation), ] Wrong continued indentation.
```

cocokale commented:
> @cocokale
> 
> > Should I put the ctcgreedysearchdecoder in another file?
> 
> Yes, I think it is the best way to handle this error (or separate text detection and text recognition - move both `ctc_greedy_search_decoder` and `ctc_beam_search_decoder` to another file)
> 
> > I've fixed all the pylint errors except the too many lines error in text_detection.py...
> 
> There is also
> 
> ```
> 14:34:25  accuracy_checker/annotation_converters/kondate.py:39: [C0330(bad-continuation), ] Wrong continued indentation.
> ```

fixed.

cocokale commented:
@eaidova Hi, there still exists some errors... Would you please show me what they are? Thanks!

eaidova commented:
@cocokale you forgot remove some imports during moving classes
```
15:44:27  ************* Module accuracy_checker.adapters.text_detection
15:44:27  accuracy_checker/adapters/text_detection.py:17: [W0611(unused-import), ] Unused defaultdict imported from collections
15:44:27  accuracy_checker/adapters/text_detection.py:24: [W0611(unused-import), ] Unused BoolField imported from config
15:44:27  accuracy_checker/adapters/text_detection.py:24: [W0611(unused-import), ] Unused ConfigError imported from config
15:44:27  accuracy_checker/adapters/text_detection.py:25: [W0611(unused-import), ] Unused CharacterRecognitionPrediction imported from representation
```

IRDonch commented:
@cocokale FYI, you can see the build errors for yourself if you click on the red cross next to the most recent commit.

cocokale commented:
> @cocokale you forgot remove some imports during moving classes
> 
> ```
> 15:44:27  ************* Module accuracy_checker.adapters.text_detection
> 15:44:27  accuracy_checker/adapters/text_detection.py:17: [W0611(unused-import), ] Unused defaultdict imported from collections
> 15:44:27  accuracy_checker/adapters/text_detection.py:24: [W0611(unused-import), ] Unused BoolField imported from config
> 15:44:27  accuracy_checker/adapters/text_detection.py:24: [W0611(unused-import), ] Unused ConfigError imported from config
> 15:44:27  accuracy_checker/adapters/text_detection.py:25: [W0611(unused-import), ] Unused CharacterRecognitionPrediction imported from representation
> ```

Sorry, my mistake. It's fixed now. :)

cocokale commented:
> @cocokale FYI, you can see the build errors for yourself if you click on the red cross next to the most recent commit.

Hi, thanks for the tip. I have clicked those links a few times before. But probably due to network issues, the website shows Network Error (tcp_error).

cocokale commented:
> From my side, that is all,
> thanks!

Thanks for the careful review. :)

cocokale commented:
I think the failure of Documentation build is because I have not update [index.md](https://github.com/opencv/open_model_zoo/blob/develop/models/intel/index.md).
In the Text Recognition section, there are two columns to fill in, namely the Complexity (GFLOPs) and Size (Mp). I'm not sure about how I should compute them... Is there a common method or benchmark? Would anyone please me some advice here? Many thanks!

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
@cocokale we got these numbers from internal infrastructure. I've updated index.md for you

cocokale commented:
> @cocokale we got these numbers from internal infrastructure. I've updated index.md for you

Thanks!

IRDonch commented:
@cocokale You didn't answer my question from before.

cocokale commented:
> @cocokale You didn't answer my question from before.

Hi, IRDonch. Sorry for the late reply. I've not taken a deep look into the C++ code yet. But I think this may take some extra time to enable this C++ demo. If we must enable this, are we able to catch the 2020.2 release?

IRDonch commented:
@cocokale It's not required per se, I'm just saying that it might be simpler to adapt the existing C++ demo than to add a whole other demo. I suggest you at least look into what that would entail while we review your Python demo.

cocokale commented:
> @cocokale It's not required per se, I'm just saying that it might be simpler to adapt the existing C++ demo than to add a whole other demo. I suggest you at least look into what that would entail while we review your Python demo.

Thanks for your suggestion. I'll look into the text_detection_demo code.

IRDonch commented on ci/requirements-ac-test.txt:
```diff
@@ -24,3 +24,4 @@ tqdm==4.42.1
 wcwidth==0.1.8            # via pytest
 yamlloader==0.5.5
 zipp==1.1.0               # via importlib-metadata
+editdistance==0.5.3
```
Please use the `update-requirements.py` script to update this file.

cocokale replied:
Hi, thanks for your careful review. I'm not sure about how to use the script to update this file. I've run this script. It didn't seem to do anything... Would you please explain a little bit more? Thanks!

IRDonch replied:
Did you get any output from the script?

cocokale replied:
(base) upsquared@UP2:~/open_model_zoo/ci$ ~/anaconda3/envs/py35_env/bin/python update-requirements.py
updating ci/requirements-ac.txt...
/home/upsquared/anaconda3/envs/py35_env/bin/python: No module named piptools
Traceback (most recent call last):
  File "update-requirements.py", line 58, in <module>
    main()
  File "update-requirements.py", line 46, in main
    'tools/accuracy_checker/requirements.in')
  File "update-requirements.py", line 34, in pip_compile
    check=True, cwd=str(repo_root))
  File "/home/upsquared/anaconda3/envs/py35_env/lib/python3.5/subprocess.py", line 398, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/home/upsquared/anaconda3/envs/py35_env/bin/python', '-mpiptools', 'compile', '--no-header', '--upgrade', '--quiet', '-o', 'ci/requirements-ac.txt', '--', 'tools/accuracy_checker/requirements.in']' returned non-zero exit status 1

IRDonch replied:
You need to install the `pip-tools` package, preferably version 4.4.1.

cocokale replied:
Hi, after installing the pip-tools==4.4.1. The script can run successfully. However, after the running, not only the editdistance has been added, some other packages has been changed and updated. Like this:
-nibabel==3.0.1
-numpy==1.17.5             # via nibabel, scikit-learn, scipy
-packaging==20.1           # via pytest
+nibabel==3.0.2
+numpy==1.18.1             # via nibabel, scikit-learn, scipy
+packaging==20.3           # via pytest
Are such modifications expected? Would you please tell me what should I do about it? Thanks!

cocokale replied:
If such modifications are not allowed, could you please help me run the script `update-requirements.py` on your side? Because in my anaconda python3.5 environment, such extra modifications will always happen when I run this script... Many thanks!

cocokale replied:
Also, there's always an HTTPSConnectionPool due to the network problem...

IRDonch replied:
> Are such modifications expected? Would you please tell me what should I do about it? Thanks!

It should be fine, although if you want to stay on the safe side, you can revert the changes, temporarily edit the script to remove the string `'--upgrade'` and rerun it. This will prevent unrelated packages from being upgraded.

IRDonch commented on demos/README.md:
```diff
@@ -32,6 +32,7 @@ The Open Model Zoo includes the following demos:
 - [Super Resolution C++ Demo](./super_resolution_demo/README.md) - Super Resolution demo (the demo supports only images as inputs). It enhances the resolution of the input image.
 - [Text Detection C++ Demo](./text_detection_demo/README.md) - Text Detection demo. It detects and recognizes multi-oriented scene text on an input image and puts a bounding box around detected area.
 - [Text Spotting Python* Demo](./python_demos/text_spotting_demo/README.md) - The demo demonstrates how to run Text Spotting models.
+- [Handwritten Japanese Recognition Python* Demo](./python_demos/handwritten_japanese_recognition_demo//README.md) - The demo demonstrates how to run Handwritten Japanese Recognition models.
```
```suggestion
- [Handwritten Japanese Recognition Python* Demo](./python_demos/handwritten_japanese_recognition_demo/README.md) - The demo demonstrates how to run Handwritten Japanese Recognition models.
```

IRDonch commented on demos/README.md:
```diff
@@ -87,6 +88,7 @@ The table below shows the correlation between models, demos, and supported plugi
 | text-detection-0003                              | [Text Detection Demo](./text_detection_demo/README.md)                                | Supported | Supported |             | Supported       |
 | text-detection-0004                              | [Text Detection Demo](./text_detection_demo/README.md)                                | Supported | Supported |             | Supported       |
 | text-recognition-0012                            | [Text Detection Demo](./text_detection_demo/README.md)                                | Supported | Supported |             |                 |
+| handwritten-japanese-recognition-0001            | [Handwritten Japanese Recognition Python* Demo](./python_demos/handwritten_japanese_recognition_demo//README.md)                             | Supported        |        Supported     |        Supported     |                 |
```
```suggestion
| handwritten-japanese-recognition-0001            | [Handwritten Japanese Recognition Python* Demo](./python_demos/handwritten_japanese_recognition_demo/README.md)                             | Supported        |        Supported     |        Supported     |                 |
```

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/README.md:
```diff
@@ -0,0 +1,82 @@
+
+The demo depends on:
+- OpenVINO library (2019R3.1 or newer)
+- Python (3.5+, which is supported by OpenVINO)
```
OMZ is a part of the OpenVINO toolkit, so it's unnecessary to specify these dependencies.

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,127 @@
+    with open(args.charlist, 'r', encoding='utf-8') as f:
+        for line in f.readlines():
+            line = line.strip('\n')
+            characters += line
```
```suggestion
        return ''.join(line.strip('\n') for line in f)
```

eaidova commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

cocokale commented:
I‘ve integrated this model into the C++ text_detection_demo. However, it seems that the opencv's function cv::putText does not support utf-8 according to [this post](https://answers.opencv.org/question/86997/how-can-i-put-text-on-image-using-opencv-puttext-function/). On my side, the result image shows "??????". Would anyone please give me some advice here? 
@IRDonch @eaidova @vladimir-dudnik

cocokale commented:
I've printed the correct prediction result with this model using the text_detection_demo. The current problem is that cv::puttext does not support non-ASCII characters, so it will display "?????" instead of correct prediction. I find that using the [freetype in opencv_contrib](https://github.com/opencv/opencv_contrib/blob/master/modules/freetype/include/opencv2/freetype.hpp#L54) could solve this problem by supporting drawing unicode characters. However, this will introduce an extra dependency opencv_contrib, which I'm not sure whether it is accepted. So, I'm looking forward to your suggestions on this issue. Many thanks!

vladimir-dudnik commented:
@dkurt @alalek could you please advice how to use cv::putText with utf-8 symbols (if that is possible at all)?

alalek commented:
OpenCV's putText() doesn't support Unicode symbols - this limitation is defined by **builtin** fonts.

Consider dumping Unicode symbols on text console (`cout` / `printf`) or use prepared pre-rendered symbol sprites (it is more complex task).

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/README.md:
```diff
@@ -0,0 +1,76 @@
+
+## How It Works
+
+The demo expects the followings model in the Intermediate Representation (IR) format:
```
```suggestion
The demo expects the following model in the Intermediate Representation (IR) format:
```

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,116 @@
+                      help='Show this help message and exit.')
+    args.add_argument("-m", "--model", type=str, required=True,
+                      help="Path to an .xml file with a trained model.")
+    args.add_argument("-i", "--input", type=str, nargs="+", required=False,
```
Shouldn't this be `required=True`? It's the input to the demo.

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,142 @@
+    args.add_argument("-lang", "--lang", type=str, default='jp',
+                      help="Language supported, default value is Japanese")
+    args.add_argument("-b", "--batchsize", default=1, type=int, help="Batch size for inference")
+    args.add_argument("-cl", "--charlist", type=str, default="./kondate_char_list.txt", help="Path to the decoding char list file")
```
You didn't fix it.

cocokale replied:
Hi, I change the path from "./kondate_char_list.txt" to "data/kondate_char_list.txt". Could you please give an example of what the path will be like? Thanks!

IRDonch replied:
Something like `os.path.join(os.path.dirname(__file__), "data/kondate_char_list.txt")`.

IRDonch commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,142 @@
+    codec = CTCCodec(characters) # in main function
+
+    # Read and pre-process input images (NOTE: one image only)
+    input_image = preprocess_input(args.input[0], height=args.input_height, width=args.input_width)[None,:,:,:]
```
You didn't fix this one.

cocokale replied:
The demo supports only image at a time, which means the -i followed by one image path.

IRDonch commented:
@cocokale OK, it looks like it'll be complicated to support Unicode characters in the C++ demo, so let's just abandon this plan. Please revert your changes to the C++ demo.

cocokale commented:
> @cocokale OK, it looks like it'll be complicated to support Unicode characters in the C++ demo, so let's just abandon this plan. Please revert your changes to the C++ demo.

In the latest commit, I used cout to show the prediction result in the terminal, so such a workaround will not be accepted here?

cocokale commented on demos/python_demos/handwritten_japanese_recognition_demo/handwritten_japanese_recognition_demo.py:
```diff
@@ -0,0 +1,116 @@
+    net = IENetwork(model=model_xml, weights=model_bin)
+
+    assert len(net.inputs) == 1, "Demo supports only single input topologies"
+    assert len(net.outputs) == 1, "Demo supports only single output topologies"
```
Hi. Thanks for the careful review. Assertion between charlist length and output shape is added. The one channel and one batch size has already been assured by line 92:
` input_image = preprocess_input(args.input[0], height=input_height, width=input_width)[None,:,:,:]
`   
Is the assertion necessary here?

Wovchena replied:
I'm not sure I understand how it assures that. `preprocess_input()` reads an image as gray scale, but if one provides a net that accepts BGR, the demo will try to put one channel image into an input with tree channels. So I suggest checking that number of channels for net's input is 1.

The similar thing with batch: if the model provided to IR has batch not 1, the demo will try to fill (batch,:,:,:) tensor with (1,:,:,:) tensor.

BTW, `nargs="+", ` should be removed from
```py
    args.add_argument("-i", "--input", type=str, nargs="+", required=True,
                      help="Required. Path to an image to infer")
```
because the demo expects a single image.

IRDonch commented:
> In the latest commit, I used cout to show the prediction result in the terminal, so such a workaround will not be accepted here?

It's very difficult to print Unicode text from C++ in a platform-independent way, so I don't think we should bother.

cocokale commented:
> > In the latest commit, I used cout to show the prediction result in the terminal, so such a workaround will not be accepted here?
> 
> It's very difficult to print Unicode text from C++ in a platform-independent way, so I don't think we should bother.

I see. Thanks for your reply.
